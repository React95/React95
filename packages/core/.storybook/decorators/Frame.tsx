import React from 'react';
import '../../components/GlobalStyle/GlobalStyle.css';
import * as themes from '../../components/themes/all';

const Frame = (Story, { globals }) => {
  const { selectedTheme } = globals;

  const theme = themes[selectedTheme] || themes.win95;

  React.useEffect(() => {
    document.body.classList.add(theme);

    return () => {
      document.body.classList.remove(theme);
    };
  }, [theme]);

  return (
    <div style={{ padding: 10 }}>
      <Story />
    </div>
  );
};

export default Frame;
