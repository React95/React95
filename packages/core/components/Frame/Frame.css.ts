import { createRainbowSprinkles, defineProperties } from 'rainbow-sprinkles';
import {
  background,
  borderRadius,
  borders,
  colors,
  displayAndBoxModel,
  font,
  outline,
  positioning,
  shadows,
  text,
  zIndices,
} from './props';
import type { SprinklesFn } from 'rainbow-sprinkles/createRuntimeFn';

const displayAndBoxModelAtomicProps = defineProperties({
  dynamicProperties: displayAndBoxModel,
  shorthands: {
    size: ['height', 'width'],
    h: ['height'],
    w: ['width'],
    minH: ['minHeight'],
    minW: ['minWidth'],

    m: ['margin'],
    mr: ['marginRight'],
    ml: ['marginLeft'],
    mt: ['marginTop'],
    mb: ['marginBottom'],
    marginX: ['marginLeft', 'marginRight'],
    marginY: ['marginTop', 'marginBottom'],
    mx: ['marginLeft', 'marginRight'],
    my: ['marginTop', 'marginBottom'],

    p: ['padding'],
    pr: ['paddingRight'],
    pl: ['paddingLeft'],
    pt: ['paddingTop'],
    pb: ['paddingBottom'],
    paddingX: ['paddingLeft', 'paddingRight'],
    paddingY: ['paddingTop', 'paddingBottom'],
    px: ['paddingLeft', 'paddingRight'],
    py: ['paddingTop', 'paddingBottom'],
  },
});

const positioningAtomicProps = defineProperties({
  dynamicProperties: positioning,
});

const colorsAtomicProps = defineProperties({
  dynamicProperties: colors,
  shorthands: {
    bgColor: ['backgroundColor'],
  },
});

const backgroundAtomicProps = defineProperties({
  dynamicProperties: background,
  shorthands: {
    bg: ['background'],
  },
});

const bordersAtomicProps = defineProperties({
  dynamicProperties: borders,
});

const borderRadiusAtomicProps = defineProperties({
  dynamicProperties: borderRadius,
});

const shadowsAtomicProps = defineProperties({
  dynamicProperties: shadows,
});

const outlineAtomicProps = defineProperties({
  dynamicProperties: outline,
});

const fontAtomicProps = defineProperties({
  dynamicProperties: font,
});

const textAtomicProps = defineProperties({
  dynamicProperties: text,
});

const zIndicesAtomicProps = defineProperties({
  dynamicProperties: zIndices,
});

export const sprinkles: SprinklesFn<
  [
    typeof colorsAtomicProps,
    typeof shadowsAtomicProps,
    typeof zIndicesAtomicProps,
    typeof displayAndBoxModelAtomicProps,
    typeof positioningAtomicProps,
    typeof backgroundAtomicProps,
    typeof bordersAtomicProps,
    typeof borderRadiusAtomicProps,
    typeof outlineAtomicProps,
    typeof fontAtomicProps,
    typeof textAtomicProps,
  ]
> = createRainbowSprinkles(
  colorsAtomicProps,
  shadowsAtomicProps,
  zIndicesAtomicProps,
  displayAndBoxModelAtomicProps,
  positioningAtomicProps,
  backgroundAtomicProps,
  bordersAtomicProps,
  borderRadiusAtomicProps,
  outlineAtomicProps,
  fontAtomicProps,
  textAtomicProps,
);

export type Sprinkles = Parameters<typeof sprinkles>[0];
