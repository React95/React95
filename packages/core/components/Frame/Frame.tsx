import React, {
  ComponentPropsWithRef,
  ElementType,
  ForwardedRef,
  forwardRef,
} from 'react';
import { Sprinkles, sprinkles } from './sprinkles.css';
import cn from 'classnames';

type FixedForwardRef = <T, P = object>(
  render: (props: P, ref: React.Ref<T>) => React.ReactNode,
) => (props: P & React.RefAttributes<T>) => React.ReactNode;

const fixedForwardRef = forwardRef as FixedForwardRef;

type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

export type FrameProps<TAs extends ElementType = 'div'> = {
  as?: TAs;
} & DistributiveOmit<
  ComponentPropsWithRef<ElementType extends TAs ? 'div' : TAs>,
  'as'
> &
  Sprinkles;

const FrameComponent = <TAs extends ElementType>(
  props: FrameProps<TAs>,
  ref: ForwardedRef<any>,
) => {
  const { as, children, ...rest } = props;
  const Component = as || 'div';
  const { className, style, otherProps } = sprinkles(rest);

  return (
    <Component
      {...otherProps}
      style={{ ...style, ...otherProps.style }}
      className={cn(className, otherProps.className)}
      ref={ref}
    >
      {children}
    </Component>
  );
};

export const Frame = fixedForwardRef(FrameComponent);
