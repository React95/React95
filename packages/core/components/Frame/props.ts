import { contract } from '../themes/contract.css';

export const positioning = {
  position: true,
  top: true,
  right: true,
  bottom: true,
  left: true,
  pointerEvents: true,
} as const;

export const displayAndBoxModel = {
  display: true,
  overflow: true,
  boxSizing: true,

  grid: true,
  gridArea: true,
  gridAutoColumns: true,
  gridAutoFlow: true,
  gridAutoRows: true,
  gridColumn: true,
  gridColumnEnd: true,
  gridColumnStart: true,
  gridRow: true,
  gridRowEnd: true,
  gridRowStart: true,
  gridTemplate: true,
  gridTemplateAreas: true,
  gridTemplateColumns: true,
  gridTemplateRows: true,

  flex: true,
  flexBasis: true,
  flexDirection: true,
  flexFlow: true,
  flexGrow: true,
  flexShrink: true,
  flexWrap: true,

  justifyContent: true,
  justifyItems: true,
  justifySelf: true,
  alignContent: true,
  alignItems: true,
  alignSelf: true,

  gap: contract.space,

  width: contract.space,
  maxWidth: contract.space,
  minWidth: contract.space,
  height: contract.space,
  maxHeight: contract.space,
  minHeight: contract.space,

  margin: contract.space,
  marginBlock: contract.space,
  marginBlockEnd: contract.space,
  marginBlockStart: contract.space,
  marginBottom: contract.space,
  marginInline: contract.space,
  marginInlineEnd: contract.space,
  marginInlineStart: contract.space,
  marginLeft: contract.space,
  marginRight: contract.space,
  marginTop: contract.space,

  padding: contract.space,
  paddingBlock: contract.space,
  paddingBlockEnd: contract.space,
  paddingBlockStart: contract.space,
  paddingBottom: contract.space,
  paddingInline: contract.space,
  paddingInlineEnd: contract.space,
  paddingInlineStart: contract.space,
  paddingLeft: contract.space,
  paddingRight: contract.space,
  paddingTop: contract.space,
} as const;

export const colors = {
  backgroundColor: contract.colors,
  color: contract.colors,
} as const;

export const background = {
  background: true,
  backgroundAttachment: true,
  backgroundImage: true,
  backgroundPosition: true,
  backgroundRepeat: true,
  backgroundSize: true,
  backgroundOrigin: true,
  backgroundClip: true,
} as const;

export const borders = {
  border: true,
  borderStyle: true,
  borderWidth: true,
  borderColor: contract.colors,
  borderTop: true,
  borderTopColor: contract.colors,
  borderTopStyle: true,
  borderTopWidth: true,
  borderRight: true,
  borderRightColor: contract.colors,
  borderRightStyle: true,
  borderRightWidth: true,
  borderBottom: true,
  borderBottomColor: contract.colors,
  borderBottomStyle: true,
  borderBottomWidth: true,
  borderLeft: true,
  borderLeftColor: contract.colors,
  borderLeftStyle: true,
  borderLeftWidth: true,
} as const;

export const borderRadius = {
  borderBottomLeftRadius: true,
  borderBottomRightRadius: true,
  borderImage: true,
  borderImageOutset: true,
  borderImageRepeat: true,
  borderImageSlice: true,
  borderImageSource: true,
  borderImageWidth: true,
  borderRadius: true,
  borderTopLeftRadius: true,
  borderTopRightRadius: true,
} as const;

export const shadows = {
  boxShadow: contract.shadows,
} as const;

export const outline = {
  outline: true,
  outlineColor: true,
  outlineStyle: true,
  outlineWidth: true,
} as const;

export const font = {
  font: true,
  fontFamily: true,
  fontStyle: true,
  fontSize: true,
  fontWeight: true,
  fontVariant: true,
  fontSizeAdjust: true,
} as const;

export const text = {
  color: contract.colors,
  direction: true,
  letterSpacing: true,
  lineHeight: true,
  textAlign: true,
  textDecoration: true,
  textIndent: true,
  textShadow: true,
  textTransform: true,
  verticalAlign: true,
  whiteSpace: true,
  wordSpacing: true,
  wordBreak: true,
} as const;

export const zIndices = {
  zIndex: contract.zIndices,
} as const;
