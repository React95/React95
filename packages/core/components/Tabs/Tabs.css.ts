import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';
import { recipe } from '@vanilla-extract/recipes';
import { calc } from '@vanilla-extract/css-utils';

export const navbar = style({
  display: 'flex',
  padding: '0',
  margin: '0',
  border: 'none',
  zIndex: 1,
});

export const navContainer = style({
  padding: contract.space[12],
  boxShadow: contract.shadows.out,
  backgroundColor: contract.colors.material,
});

export const tab = recipe({
  base: {
    listStyle: 'none',
    userSelect: 'none',
    paddingBlock: contract.space[3],
    paddingInline: contract.space[6],
    backgroundColor: contract.colors.material,
    marginRight: contract.space[2],
    marginLeft: contract.space[2],
    boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
      -2px 1px 0 -1px ${contract.colors.material},
      -3px 1px 0 -1px ${contract.colors.borderLightest},
      -2px 0 0 -1px ${contract.colors.borderLightest},
      2px 1px 0 -1px ${contract.colors.borderDark},
      2px 0 0 -1px ${contract.colors.borderDarkest},
      3px 1px 0 -1px ${contract.colors.borderDarkest}`,
    selectors: {
      '&:first-child': {
        marginLeft: contract.space[4],
      },
    },
  },
  variants: {
    active: {
      true: {
        marginTop: calc.negate(contract.space[2]),
        marginLeft: calc.negate(contract.space[2]),
        marginBottom: contract.space[1],
        '&:first-child': { marginLeft: contract.space[2] },
        boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
        -2px 1px 0 -1px ${contract.colors.material},
        -0.5px 3px 0 0.5px ${contract.colors.material},
        -2px 2px 0 0 ${contract.colors.borderLightest},
        -2px 0 0 -1px ${contract.colors.borderLightest},
        1px 2px 0 0 ${contract.colors.borderDark},
        2px 0 0 -1px ${contract.colors.borderDarkest},
        2px 2px 0 0 ${contract.colors.borderDarkest}`,
      },
    },
  },
});

globalStyle(`${tab.classNames.variants.active.true} + li`, {
  marginLeft: 0,
  boxShadow: `inset 0 1px 0 0 ${contract.colors.borderLightest},
    2px 1px 0 -1px ${contract.colors.borderDark},
    2px 0 0 -1px ${contract.colors.borderDarkest},
    3px 1px 0 -1px ${contract.colors.borderDarkest}`,
});
