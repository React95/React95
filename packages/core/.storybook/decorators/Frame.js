import React from 'react';
import { GlobalStyle } from '../../components';

const Frame = Story => (
  <div style={{ padding: 10 }}>
    <GlobalStyle />
    <Story />
  </div>
);

export default Frame;
