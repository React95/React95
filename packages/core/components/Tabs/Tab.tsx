import React, { forwardRef } from 'react';

import Frame from '../Frame';
import { tab } from './Tabs.css';
import { FrameProps } from '../Frame/Frame';

export type TabProps = {
  activeTab?: string;
  title: string;
  disabled?: boolean;
  onClick?(e: React.MouseEvent): void;
} & React.HTMLProps<HTMLLIElement> &
  Omit<FrameProps<'li'>, 'as'>;

export const Tab = forwardRef<HTMLLIElement, TabProps>(
  ({ activeTab, title, ...rest }, ref) => (
    <Frame
      {...rest}
      className={tab({ active: activeTab === title })}
      ref={ref}
      as="li"
    >
      {title}
    </Frame>
  ),
);
