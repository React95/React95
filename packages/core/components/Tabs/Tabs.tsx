import React, { forwardRef } from 'react';
import cn from 'classnames';

import Frame from '../Frame';
import { Tab, TabProps } from './Tab';
import { navContainer, navbar } from './Tabs.css';
import { FrameProps } from '../Frame/Frame';

export type TabsProps = {
  defaultActiveTab?: string;
  children: React.ReactElement<TabProps> | Array<React.ReactElement<TabProps>>;

  onChange?(title: string, e: React.MouseEvent): void;
} & Omit<FrameProps<'ol'>, 'as'>;

export const Tabs = forwardRef<HTMLOListElement, TabsProps>(
  ({ children, defaultActiveTab, onChange, ...rest }, ref) => {
    const [firstTab] = React.Children.toArray(children) as Array<
      React.ReactElement<TabProps>
    >;
    const [activeTab, setActiveTab] = React.useState(
      defaultActiveTab || firstTab.props.title,
    );

    return (
      <>
        <Frame
          {...rest}
          className={cn(navbar, rest.className)}
          as="ol"
          ref={ref}
        >
          {React.Children.map(
            children,
            (child: React.ReactElement<TabProps>) => {
              const { title, disabled } = child.props;

              return (
                <Tab
                  key={title}
                  {...child.props}
                  activeTab={activeTab}
                  onClick={e => {
                    if (!disabled) {
                      if (onChange) {
                        onChange(title, e);
                      }
                      setActiveTab(title);
                    }
                  }}
                />
              );
            },
          )}
        </Frame>

        <Frame className={navContainer} width={rest.width || rest.w}>
          {React.Children.map(
            children,
            (child: React.ReactElement<TabProps>) =>
              child.props.title === activeTab && child.props.children,
          )}
        </Frame>
      </>
    );
  },
);
