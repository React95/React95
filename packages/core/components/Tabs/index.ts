import { Tab } from './Tab';
import { Tabs } from './Tabs';

export { Tab, Tabs };
