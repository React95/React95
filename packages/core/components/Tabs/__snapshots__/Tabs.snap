// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Tabs /> > Snapshot > should match snapshot 1`] = `
<div>
  <ol
    class="Tabs_navbar__fyllcd0"
  >
    <li
      class="Tabs_tab__fyllcd2 Tabs_tab_active_true__fyllcd3"
    >
      example
    </li>
  </ol>
  <div
    class="Tabs_navContainer__fyllcd1"
  >
    <p>
      Example text
    </p>
  </div>
</div>
`;
